### PR TITLE
Split big files

### PR DESCRIPTION
**Pull request by thesunlover**
added support for huge files
check the long_test.py and play with minutes & processes number to fit in the RAM available

Edit:
Procedure of the process:
1. Splits the large audio file to fingerprintable smaller pieces
2. Fingerprints them one by one.
3. Saves them as a single audio in the DataBase.

### CODE REVIEW

**Comment by worldveil** [comment]
So to be clear, this splits up large files over a certain length (ie, 3 minutes), and fingerprints each as a separate "song", yes?

**Comment by thesunlover** [comment]
ok,

for all the split files it gives a single song; as if the original file was fingerprinted
it is an imitation of single fingerprinting

in long_test.py

djv.fingerprint_with_duration_check(long_song, minutes=3, song_name="RogerPenrose1",processes=3) 

i split the file into pieces, 3 minutes each, I use 3 CPU pieces and the song is called "RogerPenrose1"

Hope that is clear enough?
i would just add removing of help-to-split-file-directory for hdd-memory-saving purpose right after 

        try:
            shutil.rmtree("./split/")
        except WindowsError:
            pass
        return sid

File: dejavu/__init__.py [hunk 1]
```diff
@@ -16,6 +32,9 @@ class Dejavu(object):
     OFFSET = 'offset'
     OFFSET_SECS = 'offset_seconds'
 
+    SPLIT_DIR = "split_dir"
+    OVERWRITE_WHEN_SPLITING = 1
```
Comment (worldveil) [inline dejavu/__init__.py]: overwrites what?

File: dejavu/__init__.py [hunk 2]
```diff
@@ -43,7 +62,7 @@ def get_fingerprinted_songs(self):
             song_name = song[self.db.FIELD_SONGNAME]
             self.songnames_set.add(song_name)
 
-    def fingerprint_directory(self, path, extensions, nprocesses=None):
+    def fingerprint_directory(self, path, extensions, nprocesses=None, splited=False, splited_song_name=""):
```
Comment (worldveil) [inline dejavu/__init__.py]: `splitted` is not great English grammar, I'd say 'splitting' or 'split' are better wordings here.

File: dejavu/__init__.py [hunk 3]
```diff
@@ -111,6 +132,61 @@ def fingerprint_file(self, filepath, song_name=None):
             self.db.set_song_fingerprinted(sid)
             self.get_fingerprinted_songs()
 
+    def fingerprint_with_duration_check(self, input_file, minutes=5, song_name=None, processes=None):
```
Comment (worldveil) [inline dejavu/__init__.py]: does `minutes` here mean that any song longer than `minutes` will be split up and read by two separate processes?

does this actually save memory? or are we just replacing N memory with one processor with N/2 memory with 2 processors (faster but still same memory usage)?

File: dejavu/__init__.py [hunk 4]
```diff
@@ -111,6 +132,61 @@ def fingerprint_file(self, filepath, song_name=None):
...
+        song_name = song_name or songname
+        # don't refingerprint already fingerprinted files
+        if song_name in self.songnames_set:
+            print "%s already fingerprinted, continuing..." % song_name
+            return
+        file_directory = os.path.dirname(input_file)
+        output_split_path = os.path.join(file_directory, self.SPLIT_DIR)
+        try:
+            os.mkdir(output_split_path)
+        except WindowsError:
```
Comment (worldveil) [inline dejavu/__init__.py]: What error would this be?

File: dejavu/__init__.py [hunk 5]
```diff
@@ -111,6 +132,61 @@ def fingerprint_file(self, filepath, song_name=None):
...
+            pass
+        output_path = os.path.join(output_split_path, song_name)
+        try:
+            os.mkdir(output_path)
+        except WindowsError:
+            pass
```
Comment (worldveil) [inline dejavu/__init__.py]: Why are you ignoring the error here with just a `pass`?

File: dejavu/__init__.py [hunk 6]
```diff
@@ -111,6 +132,61 @@ def fingerprint_file(self, filepath, song_name=None):
...
+                                    output_file]
+            #songname for the input
+            retcode = subprocess.call(convertion_command, stderr=open(os.devnull))
+            if retcode != 0:
+                raise SplitError(input_file, output_file, retcode)
+            start_offset += split_length
+            end_offset += split_length
+            end_offset = min(end_offset, duration)
+
+            # song_name = song_name or songname
```
Comment (worldveil) [inline dejavu/__init__.py]: get rid of commented out code

File: dejavu/__init__.py [hunk 7]
```diff
@@ -111,6 +132,61 @@ def fingerprint_file(self, filepath, song_name=None):
...
+            # song_name, hashes = _fingerprint_worker(output_file,
+            #                                         self.limit,
+            #                                         song_name=song_name)
+            # self.db.insert_hashes(sid, hashes)
+        self.db.set_song_fingerprinted(sid)
+        self.get_fingerprinted_songs()
+        self.fingerprint_directory(output_path, [extension], nprocesses=processes, splited=True, splited_song_name=song_name)
+        #TODO: delete files in the output_split_path after FP
```
Comment (worldveil) [inline dejavu/__init__.py]: This should probably be done for this PR.

**Comment by worldveil** [comment]
Also the binary files (large mp3s) are too large and I'd rather not increase the size of the repo with those. 

Can you instead give a public link (could even download with `urllib2` in example.py) to some copyright-free music?

File: dejavu/__init__.py (hunk 1 above)
Comment (thesunlover) [inline dejavu/__init__.py]: its about overwriting the temp files that were split with ffmpeg.
probably wouldnt be needed if we delete the temporary directory.

File: dejavu/__init__.py (hunk 2 above)
Comment (thesunlover) [inline dejavu/__init__.py]: yes, I agree to the note.
probably : threat_as_split_from_big_file would be even better

File: dejavu/__init__.py (hunk 3 above)
Comment (thesunlover) [inline dejavu/__init__.py]: well the function that is called does the following:
first it does the splitting (yes if song is larger than minutes it splits it else it fingerprints it as a single one)
it save RAM, not HDD memory.
if we start fingerprinting the 60 minutes mp3 it raises MemoryError around 15-30 minute ...

File: dejavu/__init__.py (hunk 4 above)
Comment (thesunlover) [inline dejavu/__init__.py]: directory already exists

File: dejavu/__init__.py (hunk 5 above)
Comment (thesunlover) [inline dejavu/__init__.py]: directory already exists

Comment (thesunlover) [inline dejavu/__init__.py]: nothing needed to be done if its already there

**Comment by thesunlover** [comment]
ok, will do it later in the evening

File: dejavu/__init__.py (hunk 7 above)
Comment (thesunlover) [inline dejavu/__init__.py]: ok, will add it in the evening

File: dejavu/__init__.py (hunk 2 above)
Comment (thesunlover) [inline dejavu/__init__.py]: will change it to  "threat_as_split"

File: dejavu/__init__.py (hunk 5 above)
Comment (thesunlover) [inline dejavu/__init__.py]: actually i can remove the tree and recreate it again

**Comment by thesunlover** [comment]
updated the pull request
added auto-creation of the long file with "ffmpeg concat"
please review once again

File: dejavu/__init__.py (hunk 1 above)
Comment (thesunlover) [inline dejavu/__init__.py]: probably should have coded it as "always overwrite" with no conditions

**Comment by thesunlover** [comment]
do we need to test with other formats like ogg & flv?

**Comment by worldveil** [comment]
If pydub handles them, I think it should be fine. 

Will review this sometime this weekend I hope.

File: dejavu/__init__.py (hunk 1 above)
Comment (thesunlover) [inline dejavu/__init__.py]: renamed the constant to OVERWRITE_TEMP_FILES_WHEN_SPLITING

**Comment by thesunlover** [comment]
I have not thought about the following.
If I set the maximum audio length for straight fingerprinting to be 1 minute and use all the available CPU cores it might be faster to fingerprint even short songs. - full CPU usage and less amount of memory usage at the same time.

What do you say, Will?

File: dejavu/__init__.py (hunk 2 above)
Comment (thesunlover) [inline dejavu/__init__.py]: fixed the spelling error
variable is now "treat_as_split"

File: dejavu/__init__.py (hunk 4 above)
Comment (thesunlover) [inline dejavu/__init__.py]: i will check for an OS-independant solution with better handling
such as described here: http://stackoverflow.com/questions/17619079/python-why-does-os-makedirs-cause-windowserror

File: dejavu/__init__.py (hunk 5 above)
Comment (thesunlover) [inline dejavu/__init__.py]: probably not a good idea when fingerprinting two parallel long files...

Comment (thesunlover) [inline dejavu/__init__.py]: ah, ...
https://docs.python.org/2/library/os.html#os.makedirs says it creates the directory recursively, so will join those too calls to os.mkdir

**Comment by thesunlover** [comment]
in the last 2 commits i moved a few arguments to be properties of Dejavu.
Should I move them in the config file?

**Comment by worldveil** [comment]
will review this week - apologies for the delay!

**Comment by worldveil** [comment]
OK, I'm pulling this for review right now. 

First things first, we need to [remove the large binary files](http://git-scm.com/book/en/v2/Git-Internals-Maintenance-and-Data-Recovery#Removing-Objects) from the git history. You removed them from current version, but they are still in history (`.git` folder is 89MB).

Second, yes the arguments for fingerprinting and splitting large files should be in the config. And we should document those options in `README.md` as well.

**Comment by thesunlover** [comment]
I think it would be easier for me to 
start a new branch,
copy the changes in there
close this pull request 
and reopen it with the new branch

is that ok?

**Comment by worldveil** [comment]
sounds fine to me.

**Comment by worldveil** [comment]
were you able to create a new PR with this? would gladly merge.

**Comment by thesunlover** [comment]
Hello, worldveil
I hope I have enough time in the evening to do it at home

**Comment by thesunlover** [comment]
reposted the PR in here:
https://github.com/worldveil/dejavu/pull/87

**Comment by thesunlover** [comment]
@worldveil would you please review the new PR #87 and give advise how to properly calculate the offset_seconds for the parts that follow the first one ?

**Comment by NathanielCustom** [comment]
@thesunlover I left a comment in PR #87 on how I got offset_seconds working.

**Pull request closed without merging**